### PR TITLE
feat: object-syntax event handlers

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

https://github.com/unjs/h3/pull/417
resolves https://github.com/unjs/h3/issues/424

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This supports object-syntax event handlers with the following syntax:

```ts
defineEventHandler({
  handler: () => {},
  before: [],
  after: []
})
```

Helper validate functions need to be drafted that type-hint `event` and infer a return type. (I would suggest we also suggest `validate: { body: () => {}, query: () => {} }` which I think will be easier to directly get a type from as user can just return the validated body/query, which we can then flow through into rest of handlers.)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/event/utils.ts [hunk 1]
```diff
@@ -25,10 +36,34 @@ export function defineEventHandler<
...
+    : EventHandlerRequest,
+>(
+  handler: EventHandlerInput<Request, Response, Validator, ValidatedRequest>,
+): EventHandler<Request, Response> {
+  if (typeof handler === "function") {
+    return Object.assign(handler, { __is_handler__: true });
+  }
+  const wrapper: EventHandler<Request, any> = async (event) => {
+    if (handler.validate) {
+      await validateData(event, handler.validate);
```
Comment (pi0) [inline src/event/utils.ts:54]: Let's not include validate in first iteration. I am worried (even if small, this adds to the bundle when not validation is needed). We can rethinking about adding more proper in next iterations 👍🏼

File: src/event/utils.ts [hunk 2]
```diff
@@ -25,10 +36,34 @@ export function defineEventHandler<
...
+    }
+    for (const hook of handler.before || []) {
+      await hook(event);
```
Comment (pi0) [inline src/event/utils.ts:57]: We need to stop search when first handler sets `event.handled`

File: src/event/utils.ts [hunk 3]
```diff
@@ -25,10 +36,34 @@ export function defineEventHandler<
...
+  handler: EventHandlerInput<Request, Response, Validator, ValidatedRequest>,
+): EventHandler<Request, Response> {
+  if (typeof handler === "function") {
+    return Object.assign(handler, { __is_handler__: true });
+  }
+  const wrapper: EventHandler<Request, any> = async (event) => {
+    if (handler.validate) {
+      await validateData(event, handler.validate);
+    }
+    for (const hook of handler.before || []) {
```
Comment (pi0) [inline src/event/utils.ts:56]: Tiny perf: Maybe we can initiate iterator when `handler.before` is defined?

File: src/event/utils.ts [hunk 4]
```diff
@@ -25,10 +36,34 @@ export function defineEventHandler<
...
+      await hook(event);
+    }
+    const result = await handler.handler?.(event);
+    for (const hook of handler.after || []) {
```
Comment (pi0) [inline src/event/utils.ts:60]: Tiny perf: We can initiate `after` also when defined

File: src/event/utils.ts [hunk 5]
```diff
@@ -25,10 +36,34 @@ export function defineEventHandler<
...
+      await hook(event);
```
Comment (pi0) [inline src/event/utils.ts:61]: Would be nice to also pass `, { body }` to the hooks to intercept the response. (body, to be consistent with global hooks)

File: src/event/utils.ts (hunk 1 above)
Comment (danielroe) [reply src/event/utils.ts:54]: Makes sense! Mainly I wanted to confirm that the types would work with inferring typed request _from_ a provided validator function.

**Review by pi0 — APPROVED** [review]

**Pull request merged**